test: migrate tests/k8s/tests/03-l7-stresstest.sh

ianvernon opened the pull request:
* Migrate bash-script test to Ginkgo framework.
* Add CreateResource and DeleteResource helper functions.

Signed-off by: Ian Vernon <ian@cilium.io>

Related-to: #1841

Commit:
test: migrate tests/k8s/tests/03-l7-stresstest.sh

* Migrate bash-script test to Ginkgo framework.
* Add CreateResource and DeleteResource helper functions.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+	var path string
+
+	initialize := func() {
+		if initialized == true {
```
@eloycoto I've seen that you've written something like this in a previous PR and I've suggested to use `sync.Once`. 
@ianvernon @eloycoto Why don't we use `BeforeSuite` instead of `BeforeEach` for this test?

eloycoto replied:
@aanm Yes, but I didn't migrate all the initialize. Sync.Once works perfect. 

The `beforeSuite` is used, but for creating all the needed vms.

aanm replied:
@eloycoto but can't we have multiple before suites?

ianvernon replied:
Good idea! I created an enhancement issue to track making this change across the tests :) https://github.com/cilium/cilium/issues/2500 

I'll add this as part of my next upload.

ianvernon replied:
Fixed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+			Expect(returnCode).Should(Equal("403"), fmt.Sprintf("Unexpected connection between frontend and backend; wanted HTTP 403, got: HTTP %s", returnCode))
+
+			/*
+				//TODO -- add this ?
```
~Yes, but leave it commented along with the comment it was on the previous test:~
```
  # FIXME: Due proxy constrains (memory?) it's impossible to execute the test
  # with 1000000 requests and 200 parallel connections. It was tested with
  # 1 request and 1 parallel connection with no success.
  #
  #kubectl exec -n qa -i ${frontend_pod} -- ab -r -n 1000000 -c 200 -s 60 -v 1 "http://${backend_svc_ip}:80/"
```

No, because this should be on the nightly tests

ianvernon replied:
Comment removed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+
+		// Set debug mode to false for both cilium pods.
+		for _, ciliumPod := range ciliumPods {
+			out, err := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, "cilium config Debug=false")
```
This doesn't work as you're expecting, check #2496

ianvernon replied:
This directly is ported from the bash-script-based tests:
```
CILIUM_PODS=$(kubectl -n kube-system get pods -l k8s-app=cilium | grep cilium- | awk '{print $1}')

for pod in $CILIUM_PODS; do
    kubectl -n kube-system exec $pod -- cilium config Debug=false
done
```

aanm replied:
Ok, when I wrote the test I thought it did one thing, I confirm it's no longer the case that's why I created issue #2496.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+		defer kubectl.DeleteResource(namespace, developmentNs)
+		res.ExpectSuccess()
+
+		resources := []string{"1-frontend.json", "2-backend-server.json", "3-backend.json"}
```
Where are those files?

ianvernon replied:
*facepalm*, coming soon...

ianvernon replied:
Fixed!

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+		log.Infof("%s", res.GetStdOut())
+
+		pods, err := kubectl.WaitForServiceEndpoints(developmentNs, "", "backend", "80", helpers.HelperTimeout)
+		Expect(pods).Should(BeTrue())
```
Shouldn't this line be swapped with the next one? If there's an error, pods will be false and will we still know which error was it?

ianvernon replied:
Fixed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+		Expect(pods).Should(BeTrue())
+		Expect(err).Should(BeNil())
+
+		By("Getting information about pods in qa namespace")
```
Out of curiosity, why are you printing this (from lines 426 to lines 436) again? It was printing on line 406

ianvernon replied:
Agreed, output was too verbose. Kept first instance of these.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+
+		Expect(returnCode).Should(Equal("200"), "Unable to connect between %s and %s:80/", frontendPod, backendSvcIP)
+
+		l7StressTest := func() {
```
Unnecessary function?

ianvernon replied:
Not unnecessary; the defer call within it ensures that all policies are deleted within Cilium at the end of the function.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+				Expect(err).Should(BeNil(), fmt.Sprintf("Error deleting resource %s: %s", policyPath, err))
+
+				By("Checking that all policies were deleted in Cilium")
+				output, err := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPodK8s1, policyCmd)
```
you should also check in `ciliumPodK8s2`, no?

ianvernon replied:
Fixed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+
+		l7StressTest()
+
+		crossNamespaceTest := func() {
```
Unnecessary function as well?

ianvernon replied:
Not unnecessary; the defer call within it ensures that all policies are deleted within Cilium at the end of the function.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+			return baseString
+		}
+
+		testConnectivity := func(backendIP, frontendPod string) {
```
Can you swap the `backend` with `frontend`? We usually test `from -> to`

ianvernon replied:
Fixed.

aanm commented on test/k8sT/Policies.go:
```diff
@@ -281,5 +281,240 @@ var _ = Describe("K8sPolicyTest", func() {
+		l7StressTest()
+
+		crossNamespaceTest := func() {
+			By("Testing policy enforcement from any namespace")
```
`Testing cilium network policy enforcement from any namespace`

ianvernon replied:
Clarified.

aanm reviewed (changes requested):
Just to make sure, I assume you also saw this comment on the `03-l7-stresstest`, right?
```
# TODO: implement in nightly builds (GH-1503)
#log "running stress tests"
#kubectl exec -n qa -i ${frontend_pod} -- wrk -t20 -c1000 -d60 "http://${backend_svc_ip}:80/"
#kubectl exec -n qa -i ${frontend_pod} -- ab -r -n 1000000 -c 200 -s 60 -v 1 "http://${backend_svc_ip}:80/"
```

Commit:
test/k8sT: address review comments

* Use sync.Once to more elegantly run initialize function only once
* Flip order of testConnectivity arguments to be more logical

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: remove TODO statement

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT/manifests: add deployment files

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: address review comments

* Make sure policies are deleted in both Cilium pods
* Make "By" messages more specific

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/helpers: remove logging of if ExecPodCmd failed

ExecPodCmd logged to console at error level if a command failed. Sometimes, an
error is expected from the provided command to the function, which emits error
logs to console. Since the function already returns an error, which tests can
decide if they want to assert being nil or not, remove the logging of the
failure of the command to console.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT: refactor common code into functions

Signed-off by: Ian Vernon <ian@cilium.io>
